Src/LiquidProjections.NHibernate/IProjectionCache.cs

```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace LiquidProjections.NHibernate
+{
+    /// <summary>
+    /// Defines a write-through cache that the <see cref="NHibernateProjector{TProjection,TKey,TState}"/> can use to avoid unnecessary loads
+    /// of the projection.
+    /// </summary>
+    public interface IProjectionCache<TProjection, in TKey>
+    {
+        /// <summary>
+        /// Adds a new item to the cache. 
+        /// </summary>
+        /// <remarks>
+        /// This method must be safe in multi-threaded scenarios.
+        /// </remarks>
+        void Add(TProjection projection);
+
+        /// <summary>
+        /// Attempts to get the item identified by <paramref name="key"/> from the cache, or creates a new one. 
+        /// </summary>
+        /// <remarks>
+        /// This method must be safe in multi-threaded scenarios. So multiple concurrent requests for the same key must always 
+        /// return the same object.
+        /// </remarks>
+        Task<TProjection> Get(TKey key, Func<Task<TProjection>> createProjection);
+
+        /// <summary>
+        /// Attempts to get the item identified by <paramref name="key"/> from the cache.
+        /// Returns <c>null</c> if the item is not in the cache.
+        /// </summary>
+        /// <remarks>
+        /// This method must be safe in multi-threaded scenarios. So multiple concurrent requests for the same key must always 
+        /// return the same object.
+        /// </remarks>
+        Task<TProjection> TryGet(TKey key);

```

Comment:
Why is this method asynchronous but `Add` and `Remove` are not?

If the cache access can be a lengthy operation in some implementation then `TryGet` should be asynchronous indeed. But in that case `Remove` at least should be asynchronous too as we want to make sure that the cache is not outdated. `Add` could still be fire-and-forget operation but better not as it should still respect the order of different operations even in that case.

If only the cache-miss handler is asynchronous then only `Get` should be asynchronous.

Reply:
The only reason why `Get` is asynchronous is because the code behind the `createProjection` delegate might be I/O intensive.